### ISSUE TITLE
add tests for synonym() linking to hybrid_property, assocationproxy, decide either if this should work or raise and get it settled

### ISSUE THREAD

**Issue by jace** [sqlalchemy/sqlalchemy#6267]
**Describe the bug**

A synonym of a hybrid property doesn't support `hasattr` calls. It raises an exception similar to the one described in #4521, but it shouldn't happen here.

**Expected behavior**
`hasattr` should not cause an exception.

**To Reproduce**

Run with pytest:

```py
from sqlalchemy import Column, Integer
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import synonym

Base = declarative_base()


class TestModel(Base):
    __tablename__ = 'test_model'

    @hybrid_property
    def foo(self):
        return "foo"

    id = Column(Integer, primary_key=True)  # NOQA: A003
    alias_foo = synonym('foo')


def test_hasattr_hybrid():
    """This test passes."""
    assert hasattr(TestModel.foo, 'test') is False


def test_use_synonym():
    """A synonym to a hybrid property can be accessed."""
    assert TestModel().foo == 'foo'
    assert TestModel().alias_foo == 'foo'


def test_hasattr_synonym():
    """This test fails."""
    assert hasattr(TestModel.alias_foo, 'test') is False
```

**sqlalchemy.exc.InvalidRequestError: synonym() attribute "TestModel.foo" only supports ORM mapped attributes, got <sqlalchemy.orm.attributes.create_proxied_attribute.<locals>.Proxy object at 0x...>**

Traceback from SQLAlchemy 1.3.24. Same error occurs in 1.4.7 but line numbers have shifted.

```
../../lib/python3.8/site-packages/sqlalchemy/orm/attributes.py:389: in __getattr__
    comparator = self.comparator
../../lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:893: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
../../lib/python3.8/site-packages/sqlalchemy/orm/attributes.py:348: in comparator
    self._comparator = self._comparator()
../../lib/python3.8/site-packages/sqlalchemy/orm/descriptor_props.py:80: in <lambda>
    lambda: self._comparator_factory(mapper),
../../lib/python3.8/site-packages/sqlalchemy/orm/descriptor_props.py:695: in _comparator_factory
    prop = self._proxied_property
../../lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:893: in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <SynonymProperty at 0x7d26513ae280; alias_foo>

    @util.memoized_property
    def _proxied_property(self):
        attr = getattr(self.parent.class_, self.name)
        if not hasattr(attr, "property") or not isinstance(
            attr.property, MapperProperty
        ):
>           raise sa_exc.InvalidRequestError(
                """synonym() attribute "%s.%s" only supports """
                """ORM mapped attributes, got %r"""
                % (self.parent.class_.__name__, self.name, attr)
            )
E           sqlalchemy.exc.InvalidRequestError: synonym() attribute "TestModel.foo" only supports ORM mapped attributes, got <sqlalchemy.orm.attributes.create_proxied_attribute.<locals>.Proxy object at 0x7d2651372450>

../../lib/python3.8/site-packages/sqlalchemy/orm/descriptor_props.py:687: InvalidRequestError
```

**Versions.**
 - OS: Linux x86_64
 - Python: 3.8.5
 - SQLAlchemy: 1.3.24 and 1.4.7
 - Database: NA
 - DBAPI: NA

**Additional context**

I stumbled onto this while working on unrelated code that examines all members of a class for markers. I want to annotate some columns as containing sensitive data like this: `sensitive_data = immutable(Column(String))`. The `immutable` wrapper leaves a tag in `Column.info`. A separate `mapper_configured` event handler looks for these markers and adds validators. The event handler blows up whenever it encounters a synonym to a hybrid property.

**Comment by zzzeek** [comment]
even though #4521 shows me doing a synonym(assocaition_proxy), there's no tests for that in test_assocaitionproxy, and similarly there are no tests for synonym(hybrid_property).  this is not something that was ever expected to work but there aren't necsesarily strong reasons it shoudlnt work.  will have to revisit the case at https://github.com/sqlalchemy/sqlalchemy/issues/4521#issuecomment-476337072 .

**Comment by zzzeek** [comment]
weird the synonym works but hasattr doesn't

**Comment by jace** [comment]
Ran into a separate issue with synonyms in 1.4. A synonym to a column can't be used in a query. It raises:

`AttributeError: Neither 'property' object nor 'Comparator' object associated with Model.synonym_name has an attribute '_extra_criteria'`.

This may be a separate issue. Trying to isolate it for a test.

**Comment by zzzeek** [comment]
> Ran into a separate issue with synonyms in 1.4. A synonym to a column can't be used in a query. It raises:
> 
> `AttributeError: Neither 'property' object nor 'Comparator' object associated with Model.synonym_name has an attribute '_extra_criteria'`.
> 
> This may be a separate issue. Trying to isolate it for a test.

taht will be fixed as part of #6272

**Comment by zzzeek** [comment]
oh that's the issue you raised, got it
